src/keycloak/chart/templates/keycloak-logging-recording-rules.yaml

```
@@ -0,0 +1,29 @@
+# Copyright 2025 Defense Unicorns
+# SPDX-License-Identifier: AGPL-3.0-or-later OR LicenseRef-Defense-Unicorns-Commercial
+
+{{- if .Values.detailedObservability }}
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: {{ include "keycloak.fullname" . }}
+  namespace: {{ .Release.Namespace }}
+  labels:
+    loki_rule: "1"
+data:
+  recording-rules.yaml: |
+    groups:
+      - name: keycloak-change-metrics
+        interval: 30s
+        rules:
+          - record: uds_keycloak:realm_modifications
+            expr: |
+              sum by (realmId) (count_over_time({namespace="keycloak"} |= "uds.keycloak.plugin.eventListeners.JSONLogEventListenerProvider" | json | eventType="ADMIN" | realmId="uds"[5m])) > 0
```

```suggestion
              sum by (realmId) (count_over_time({app="keycloak", namespace="keycloak"} | json | loggerName="uds.keycloak.plugin.eventListeners.JSONLogEventListenerProvider" | eventType="ADMIN" | label_format realmId="{{realmID}}" [30s]))
```
couple suggestions in here:
1. add label filter for keycloak pods
2. convert to json immediately and lookup `loggerName="uds.keycloak.plugin.eventListeners.JSONLogEventListenerProvider"`
3. remove `realmId="uds"` as you are already summing by realmId
4. remove >0 (make this metric an actual counting total instead of a boolean)
5. set the count over time window equal to the loki recording window (30s), this gives us a monotonic counter which is a more useful metric for both dashboards and alerting (you could also increase the loki evaluation period, it just needs to match the rolling count for a monotonic counter)
6. add the realmId as an explicit label so the sum by works (and alerting is properly grouped by label)

---

also is there a reason why Keycloak doesn't emit the master realm events? i tried trigger a log event in the master realm locally, but couldn't (only in the UDS realm)

---

The master realm logging needs to be explicitly enabled: https://uds.defenseunicorns.com/reference/configuration/single-sign-on/notifications-and-alerts/#master-realm-event-logging